Testnet/Ticketbooth/Ticketbooth/TicketContract_1_0_0.cs

```diff
@@ -0,0 +1,474 @@
+﻿using Stratis.SmartContracts;
+
+public class TicketContract_1_0_0 : SmartContract
+{
+    /// <summary>
+    /// Creates a new ticketing contract
+    /// </summary>
+    /// <param name="smartContractState"></param>
+    /// <param name="seatsBytes">The serialized array of seats</param>
+    /// <param name="venueName">The venue that hosts the contract</param>
+    public TicketContract_1_0_0(ISmartContractState smartContractState, byte[] seatsBytes, string venueName)
+        : base(smartContractState)
+    {
+        var seats = Serializer.ToArray<Seat>(seatsBytes);
+        var tickets = new Ticket[seats.Length];
+
+        for (int i = 0; i < seats.Length; i++)
+        {
+            tickets[i] = new Ticket { Seat = seats[i] };
+        }
+
+        Log(new Venue { Name = venueName });
+        Owner = Message.Sender;
+        Tickets = tickets;
+    }
+
+    /// <summary>
+    /// Stores ticket data for the contract
+    /// </summary>
+    public Ticket[] Tickets
+    {
+        get
+        {
+            return PersistentState.GetArray<Ticket>(nameof(Tickets));
+        }
+        private set
+        {
+            PersistentState.SetArray(nameof(Tickets), value);
+        }
+    }
+
+    private ulong EndOfSale
+    {
+        get
+        {
+            return PersistentState.GetUInt64(nameof(EndOfSale));
+        }
+        set
+        {
+            PersistentState.SetUInt64(nameof(EndOfSale), value);
+        }
+    }
+
+    private ulong ReleaseFee
+    {
+        get
+        {
+            return PersistentState.GetUInt64(nameof(ReleaseFee));
+        }
+        set
+        {
+            PersistentState.SetUInt64(nameof(ReleaseFee), value);
+            Log(new TicketReleaseFee { Amount = value });
+        }
+    }
+
+    private ulong NoRefundBlockCount
+    {
+        get
+        {
+            return PersistentState.GetUInt64(nameof(NoRefundBlockCount));
+        }
+        set
+        {
+            PersistentState.SetUInt64(nameof(NoRefundBlockCount), value);
+            Log(new NoRefundBlocks { Count = value });
+        }
+    }
+
+    private bool RequireIdentityVerification
+    {
+        get
+        {
+            return PersistentState.GetBool(nameof(RequireIdentityVerification));
+        }
+        set
+        {
+            PersistentState.SetBool(nameof(RequireIdentityVerification), value);
+            Log(new IdentityVerificationPolicy { RequireIdentityVerification = value });
+        }
+    }
+
+    private Address Owner
+    {
+        get
+        {
+            return PersistentState.GetAddress(nameof(Owner));
+        }
+        set
+        {
+            PersistentState.SetAddress(nameof(Owner), value);
+        }
+    }
+
+    private bool SaleOpen
+    {
+        get
+        {
+            var endOfSale = EndOfSale;
+            return endOfSale != 0 && Block.Number < endOfSale;
```

Comment:
Have you thought about the pros and cons of the ability to alter this if necessary once a sale has been started? For example, extending the sale? Not saying you need to do it, just something to think about.

Reply:
This might be a good idea to implement in a future version. I can't think of any significant drawback of allowing the sale to be extended, off the top of my head.

Reply:
Cool, yeah just thinking about real business use cases. Sales "extend" all the time, whether it was a premeditated extension prior to opening up sales or not.